draft-ietf-quic-transport.md

```diff
@@ -2427,6 +2407,32 @@ ACK Block (repeated):
   packets preceding the largest packet number, as determined by the
   preceding Gap.
 
+### Sending ACK Frames
+
+Implementations MUST NOT generate packets that only contain ACK frames in
+response to packets which only contain ACK frames. However, they SHOULD
+acknowledge packets containing only ACK frames when sending ACK frames in
+response to other packets.  Implementations MUST NOT send more than one ACK
+frame per received packet that contains frames other than ACK frames.  Packets
+containing non ACK frames MUST be acknowledged immediately or when a delayed
+ack timer expires.
+
+To limit ACK blocks to those that have not yet been received by the sender, the
+receiver SHOULD track which ACK frames have been acknowledged by its peer.  Once
```

Comment:
Could we also recommend or at least suggest tracking what acknowledgments are outstanding (sent but not yet acknowledged)?  There might be some value in not re-sending all unacknowledged acknowledgments (where ack_delay_timer << RTT especially).

Reply:
Perhaps a suggestion, but I think we should encourage receivers to re-ack as much as they can, for resilience and performance. How about something like: "In every ACK frame, a receiver SHOULD attempt to acknowledge all received packets whose acknowledgement is not yet known to have reached the sender. However, a receiver MAY avoid re-acknowledging all of them in subsequent ACK frames, and it may track outstanding acknowledgements to do so. Note that doing so reduces the connection's resilience to loss of acknowledgements." I think the wording could be improved, but something like this?

Reply:
Hm. I think this is captured in the "to limit receiver state below". Do we need any more text here? I think it may still be valuable to state the first sentence in my suggestion above perhaps?

Reply:
This isn't an example of limiting state, it's limiting bytes transmitted.  I think that your formulation here (MAY avoid sending an acknowledgment if an ACK frame containing the same acknowledgment is already in flight) is a good way to deal with this.

The limiting of state does end up having the same net effect (fewer ACK ranges), but it's because the receiver is discarding information.  This would require more effort.

Reply:
I think this limiting of state is described a few paragraphs above.

Reply:
limiting of state is below.  I'm making a different point.  We don't send stream data that we just sent, so it is possible that duplicating acknowledgments in the same way is inefficient.  Now, Jana rightly observes that the cost is low and the benefits high, which is an excellent point, but there might be cases where the cost is high enough (large BDP connections for example) to warrant trimming ACK frames down in size.

Reply:
Ok, understood.  In that case, I think the text below is sufficient for now?